lib/url/URLGenerator.js

```diff
@@ -0,0 +1,24 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+const { RawSource } = require("webpack-sources");
+
+class URLGenerator {
+	generate(module, { moduleGraph }) {
+		let source = module.originalSource().source();
+
+		const issuer = moduleGraph.getIssuer(module);
+
+		if (issuer && issuer.type.includes("javascript")) {
```

Comment:
The idea is good, but issuer is the wrong thing for that.

You can use `moduleGraph.getIncomingConnections()` to get all referencing modules (issuer is only the first one).

I would move the logic of determining if the module should be rendered to JS to `JavascriptModulesPlugin.shouldRender`. (Oh it's already there, so it can be removed here)

Split this into two generators: UrlJavascriptGenerator and UrlAssetGenerator.
One generates the JS module referencing the asset as URL.
The other emits the raw asset (just returning originalSource).

(Similar to the WASM generators)